staging/src/k8s.io/apiserver/pkg/admission/cel/metrics.go

```diff
@@ -0,0 +1,104 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cel
+
+import (
+	"context"
+	"time"
+
+	"k8s.io/component-base/metrics"
+	"k8s.io/component-base/metrics/legacyregistry"
+)
+
+const (
+	metricsNamespace = "apiserver"
+	metricsSubsystem = "validating_admission_policy"
+)
+
+var (
+	// Metrics provides access to validation admission metrics.
+	Metrics = newValidationAdmissionMetrics()
+)
+
+// ValidatingAdmissionPolicyMetrics aggregates Prometheus metrics related to validation admission control.
+type ValidatingAdmissionPolicyMetrics struct {
+	policyCheck      *metrics.CounterVec
+	policyDefinition *metrics.CounterVec
+	policyLatency    *metrics.HistogramVec
+}
+
+func newValidationAdmissionMetrics() *ValidatingAdmissionPolicyMetrics {
+	check := metrics.NewCounterVec(
+		&metrics.CounterOpts{
+			Namespace:      metricsNamespace,
+			Subsystem:      metricsSubsystem,
+			Name:           "check_total",
+			Help:           "Validation admission policy check total, labeled by policy and further identified by binding, validation expression, and enforcement action taken.",
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"policy", "policy_binding", "validation_expression", "enforcement_action"},
+	)
+	definition := metrics.NewCounterVec(&metrics.CounterOpts{
+		Namespace:      metricsNamespace,
+		Subsystem:      metricsSubsystem,
+		Name:           "definition_count",
+		Help:           "Validation admission policy count total, labeled by state and enforcement action.",
+		StabilityLevel: metrics.ALPHA,
+	},
+		[]string{"state", "enforcement_action"},
+	)
+	latency := metrics.NewHistogramVec(&metrics.HistogramOpts{
+		Namespace: metricsNamespace,
+		Subsystem: metricsSubsystem,
+		Name:      "check_duration_seconds",
+		Help:      "Validation admission latency for individual validation expressions in seconds, labeled by policy and param resource, further including state and enforcement action taken.",
+		// the bucket distribution here is spaced the same as for webhook
+		// metrics, with the initial/smallest value (10ms) chosen based on the
+		// minimum time needed for a CEL expression based on the benchmarks
+		// done for CRD evaluation
+		// however, there are only 4 buckets here (compared to the webhook's 6)
+		// since the cost estimation system will prevent most expression that
+		// take longer than 1s from being used in the first place
```

Comment:
Why use buckets tuned for a remote requests for in-process evaluation? Naively I was expecting this to be at least an order of magnitude faster. Have we considered adding a bucket or two on the low end to replace those removed from the high end? Do we have any benchmarks to help guide this?

Reply:
I've added a [benchmark suite repository](https://github.com/DangerOnTheRanger/cel-benchmark) and adjusted the buckets based on the results I got:
```
goos: linux
goarch: amd64
pkg: github.com/DangerOnTheRanger/cel-benchmark
cpu: Intel(R) Xeon(R) CPU @ 2.30GHz
BenchmarkNoop-16        11726302               101.7 ns/op
BenchmarkIterate-16        95521             12326 ns/op
BenchmarkAccess-16       6559255               180.6 ns/op
PASS
ok      github.com/DangerOnTheRanger/cel-benchmark      4.027s

goos: linux
goarch: amd64
pkg: github.com/DangerOnTheRanger/cel-benchmark
cpu: Intel(R) Xeon(R) CPU @ 2.30GHz
BenchmarkNoop-16        11635830               104.7 ns/op
BenchmarkIterate-16        94740             13002 ns/op
BenchmarkAccess-16       6500756               182.9 ns/op
PASS
ok      github.com/DangerOnTheRanger/cel-benchmark      4.121s

goos: linux
goarch: amd64
pkg: github.com/DangerOnTheRanger/cel-benchmark
cpu: Intel(R) Xeon(R) CPU @ 2.30GHz
BenchmarkNoop-16        10102984               126.0 ns/op
BenchmarkIterate-16        78733             13204 ns/op
BenchmarkAccess-16       6346537               183.6 ns/op
PASS
ok      github.com/DangerOnTheRanger/cel-benchmark      4.002s
```